akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolSlot.scala

```diff
@@ -36,240 +35,133 @@ private object PoolSlot {
-          }) :: Nil
-        } else {
-          inflightRequests.map { rc ⇒
-            if (rc.retriesLeft == 0) {
-              val reason = error.fold[Throwable](new UnexpectedDisconnectException("Unexpected disconnect"))(ConstantFun.scalaIdentityFunction)
-              connInport ! ActorPublisherMessage.Cancel
-              ResponseDelivery(ResponseContext(rc, Failure(reason)))
-            } else SlotEvent.RetryRequest(rc.copy(retriesLeft = rc.retriesLeft - 1))
-          }(collection.breakOut)
+          connectionFlowSink = new SubSinkInlet[HttpResponse]("ResponseSink")
+          connectionFlowSink.setHandler(connectionInFlowHandler)
+
+          isConnected = true
         }
-      }
-      inflightRequests = immutable.Queue.empty
-      onNext(SlotEvent.Disconnected(slotIx, results.size) :: results)
-      if (canceled) onComplete()
 
-      context.become(unconnected)
-    }
+        grab(in) match {
+          case ConnectEagerlyCommand ⇒
+            if (!isConnected) {
+              createSubSourceOutlets()
 
-    override def onComplete(): Unit = {
-      exposedPublisher.shutdown(None)
-      super.onComplete()
-      shutdown()
-    }
+              Source.fromGraph(connectionFlowSource.source)
+                .via(connectionFlow).runWith(Sink.fromGraph(connectionFlowSink.sink))(subFusingMaterializer)
 
-    override def onError(cause: Throwable): Unit = {
-      exposedPublisher.shutdown(Some(cause))
-      super.onError(cause)
-      shutdown()
-    }
+              connectionFlowSink.pull()
+            }
 
-    def shutdown(): Unit = context.stop(self)
-  }
+            emit(out1, ConnectedEagerly(slotIx), () ⇒ pull(in))
 
-  private case class FromConnection(ev: Any) extends NoSerializationVerificationNeeded
-
-  private class FlowInportActor(slotProcessor: ActorRef) extends ActorPublisher[HttpRequest] with ActorLogging {
-    def receive: Receive = {
-      case ev: Request            ⇒ slotProcessor ! FromConnection(ev)
-      case OnNext(r: HttpRequest) ⇒ onNext(r)
-      case OnComplete             ⇒ onCompleteThenStop()
-      case OnError(e)             ⇒ onErrorThenStop(e)
-      case Cancel ⇒
-        slotProcessor ! FromConnection(Cancel)
-        context.stop(self)
-    }
-  }
-  def flowInportProps(s: ActorRef) = Props(new FlowInportActor(s)).withDeploy(Deploy.local)
-
-  private class FlowOutportActor(slotProcessor: ActorRef) extends ActorSubscriber with ActorLogging {
-    def requestStrategy = ZeroRequestStrategy
-    def receive: Receive = {
-      case Request(n) ⇒ request(n)
-      case Cancel     ⇒ cancel()
-      case ev: OnNext ⇒ slotProcessor ! FromConnection(ev)
-      case ev @ (OnComplete | OnError(_)) ⇒
-        slotProcessor ! FromConnection(ev)
-        context.stop(self)
+          case DispatchCommand(rc: RequestContext) ⇒
+            if (isConnected) {
+              inflightRequests.add(rc)
+              connectionFlowSource.push(rc.request)
```

Comment:
Again: I think we should signal that we don't accept further requests `if (rc.request.header[Connection].exists(_.hasClose))`.